Add support for remote COPC reading

NEDJIMAbelgacem opened the pull request:
## Description
This PR adds the necessary code to be able to read COPC files from network.
Following up on https://github.com/qgis/QGIS/pull/47932
Currently caching is disabled and will be implemented as a separate feature.

NEDJIMAbelgacem committed:
- Implement remote COPC without caching
- COPC code improvements

NEDJIMAbelgacem committed:
move lazperf point records into decodePoint

NEDJIMAbelgacem committed:
Add docs

NEDJIMAbelgacem committed:
Fix CI build

wonder-sk commented on src/core/pointcloud/qgsremotecopcpointcloudindex.h:
```diff
@@ -0,0 +1,100 @@
+#include "qgspointcloudindex.h"
+#include "qgspointcloudattribute.h"
+#include "qgsstatisticalsummary.h"
+#include "qgis_sip.h"
```
this include should not be needed

wonder-sk commented on src/core/pointcloud/qgsremotecopcpointcloudindex.cpp:
```diff
@@ -0,0 +1,373 @@
+  // Request header data
+  {
+    QNetworkRequest nr( mUrl );
+    nr.setAttribute( QNetworkRequest::HttpPipeliningAllowedAttribute, true );
```
is this required? does it have some effect here? as we're using blocking network request here, it does not seem like we would be able to put multiple requests into the pipeline?

NEDJIMAbelgacem replied:
Not really I was experimenting with that flag a bit and forgot it there

wonder-sk commented on src/core/pointcloud/qgsremotecopcpointcloudindex.cpp:
```diff
@@ -0,0 +1,373 @@
+
+  double calculatedSpan = nodeMapExtent( root() ).width() / mCopcInfoVlr.spacing;
+  mSpan = calculatedSpan;
+
```
A lot of the code here seems duplicated from the local COPC implementation. It may be useful to either move it to some utils class - or - have a copc point cloud index class that would contain members and functions that are common for both local and remote COPC index, with a class hierarchy like this:
```
   QgsPointCloudIndex
                 ^
                 |
QgsCopcPointCloudIndex
                 ^
                 |
   ---------------------------------
  |                                |
QgsLocalCopcPointCloudIndex   QgsRemoteCopcPointCloudIndex

wonder-sk commented on src/core/pointcloud/qgsremotecopcpointcloudindex.cpp:
```diff
@@ -0,0 +1,373 @@
+  QgsBlockingNetworkRequest::ErrorCode errCode = req.get( nr );
+  if ( errCode != QgsBlockingNetworkRequest::NoError )
+  {
+    QgsDebugMsg( QStringLiteral( "Request failed: " ) + mUrl.toString() );
```
we should probably return here and not rely on the fact that the reply has empty content - it could be that server sends some response anyway, and we end up writing some wrong data to the hierarchy...

wonder-sk commented on src/core/pointcloud/qgsremotecopcpointcloudindex.h:
```diff
@@ -0,0 +1,100 @@
+    lazperf::wkt_vlr mWktVlr;
+    lazperf::eb_vlr mExtraBytesVlr;
+    mutable lazperf::header14 mCopcHeader;
+    mutable lazperf::copc_info_vlr mCopcInfoVlr;
```
as discussed, it would be good to put some of these pieces of data to a struct that would be passed to decompress function, so we do not need to do redundant parsing of the same data

wonder-sk commented:
Some more feedback from testing:
- I was getting crashes when moving/zooming 2d canvas. It looks like there is a multi-threading issue - if I wait for canvas to fully finish rendering and then move/zoom, it seems to work fine. But if I do multiple moves/zooms in short period of time, it crashes.
  - it seems the issue is that you updated findEntryForRequest() and addEntry() calls in tile download manager, but not updateEntry() and removeEntry(), leaving it sometimes in invalid state, it seems this fixes the crash:
```diff
--- a/src/core/qgstiledownloadmanager.cpp
+++ b/src/core/qgstiledownloadmanager.cpp
@@ -288,7 +288,7 @@ void QgsTileDownloadManager::updateEntry( const QgsTileDownloadManager::QueueEnt
 {
   for ( auto it = mQueue.begin(); it != mQueue.end(); ++it )
   {
-    if ( entry.request.url() == it->request.url() )
+    if ( entry.request.url() == it->request.url() && entry.request.rawHeader( "Range" ) == it->request.rawHeader( "Range" ) )
     {
       *it = entry;
       return;
@@ -302,7 +302,7 @@ void QgsTileDownloadManager::removeEntry( const QNetworkRequest &request )
   int i = 0;
   for ( auto it = mQueue.constBegin(); it != mQueue.constEnd(); ++it, ++i )
   {
-    if ( it->request.url() == request.url() )
+    if ( it->request.url() == request.url() && it->request.rawHeader( "Range" ) == request.rawHeader( "Range" ) )
     {
       mQueue.removeAt( i );
       return;
```    

- I have seen a problem that after zooming in more, nothing got shown in 2D canvas and "Caught unhandled std::exception: std::bad_alloc" appeared in terminal - not sure if it's related to the previous issue

- if I use `https://s3.amazonaws.com/hobu-lidar/melbourne-2018.copc.laz` as input, the layer name is parsed as "hobu-lidar" using the logic from EPT. It would make more sense to instead name the layer "melbourne-2018" from the file's basename
- melbourne dataset - layer properties reports that "data format" is 135 - which is strange, it should be 7:
  
![image](https://user-images.githubusercontent.com/193367/161832087-15efd7af-e2cd-4fd3-935f-f1209381dd0a.png)

NEDJIMAbelgacem committed:
Split QgsPointCloudRequest into 2 classes

NEDJIMAbelgacem committed:
- Add a class for handling laz header processing
- Unify some of the remote COPC code with local COPC
- Address some of Martin's reviews

wonder-sk commented on src/core/pointcloud/qgslazinfo.cpp:
```diff
@@ -0,0 +1,300 @@
+  mIsValid = true;
+}
+
+QgsLazInfo::QgsLazInfo( QUrl &url )
```
As discussed, I would prefer not to have this class do any network requests, only process byte arrays that are provided to it. I would suggest that:
- there is just a single constructor that initialized empty class
- there is a call to parse header from byte array
- there is another call to parse VLR data & everything else

We can have some utility functions to get those byte arrays from std::ifstream or from QgsBlockingNetworkRequest.

wonder-sk commented on src/core/pointcloud/qgslazinfo.h:
```diff
@@ -0,0 +1,114 @@
+
+#define SIP_NO_FILE
+
+class CORE_EXPORT QgsLazInfo
```
Please add a docstring

wonder-sk commented on src/core/pointcloud/qgslazinfo.h:
```diff
@@ -0,0 +1,114 @@
+#define SIP_NO_FILE
+
+class CORE_EXPORT QgsLazInfo
+{
```
One thing that is confusing about this class is that most of the code is there only for COPC support - there are various assumptions that the LAZ file is 1.4 (point record 6 / 7 / 8), but then there is parseExtrabytes() function that is used also for EPT, which can contain LAZ 1.2.

My suggestion is to rename this class e.g. to `QgsCopcHeaderInfo` and keep it limited to COPC functionality. Anything that is meant for both EPT and COPC should probably go back to QgsLazDecoder - mainly parseExtrabytes(), not sure if there's anything else.

NEDJIMAbelgacem replied:
COPC is just LAZ and all things that we are doing in this class are common between the two. So instead of splitting the handling of which attributes exist in the dataset between the two, it should be unified IMO. The only important thing that COPC has over standard LAZ is a VLR that points to the hierarchy root and that is a relatively small portion of code that can be handled in the point cloud index.
The way we handle EPT header info feels a bit complicated because we never actually read the original file and we don't parse its header data (I mean the original laz not ept.json). What we are doing with untwine in a sense is just splitting the original dataset file into many laz files and in my understanding the header data that is taken from the ept.json file shouldn't be used to read the tiles in the first place. EPT is in kinda agnostic to the tiles format (the individual tiles could be binary, zstd or laz). I am suggesting that we keep it as QgsLazInfo and this class should be used for decoding LAZ data because we actually won't need any parameters from EPT to read a LAZ tile correctly.

Sorry for the long comment 😅 . We can discuss it in more details tommorow and I will probably have a better code for this :)

wonder-sk commented on src/core/pointcloud/qgslazdecoder.h:
```diff
@@ -23,6 +23,7 @@
 #include "qgis_sip.h"
 #include "qgspointcloudblock.h"
 #include "qgspointcloudattribute.h"
+#include "qgslazinfo.h"
```
forward declaration of QgsLazInfo should be enough, without #include

wonder-sk commented on src/core/pointcloud/qgsremotecopcpointcloudindex.cpp:
```diff
@@ -0,0 +1,211 @@
+/***************************************************************************
+                         qgsremoteeptpointcloudindex.cpp
```
ept -> copc

wonder-sk commented on src/core/pointcloud/qgsremotecopcpointcloudindex.cpp:
```diff
@@ -0,0 +1,211 @@
+                         qgsremoteeptpointcloudindex.cpp
+                         --------------------
+    begin                : March 2021
+    copyright            : (C) 2021 by Belgacem Nedjima
```
2022 ;-)

NEDJIMAbelgacem commented on src/core/pointcloud/qgscopcpointcloudblockrequest.h:
```diff
@@ -0,0 +1,64 @@
+    int mPointCount;
+    QgsLazInfo mLazInfo;
+  protected:
+    void blockFinishedLoading();
```
Yeah that's a bit surprising but I added "slot" just in case :)

NEDJIMAbelgacem committed:
- Refactor QgsLazInfo
- Address some reviews

NEDJIMAbelgacem commented on src/core/pointcloud/qgscopcpointcloudblockrequest.cpp:
```diff
@@ -0,0 +1,80 @@
+    const QgsVector3D &scale, const QgsVector3D &offset, const QgsPointCloudExpression &filterExpression,
+    uint64_t blockOffset, int32_t blockSize, int pointCount, const QgsLazInfo &lazInfo )
+  : QgsPointCloudBlockRequest( node, uri, attributes, requestedAttributes, scale, offset, filterExpression ),
+    mBlockOffset( blockOffset ), mBlockSize( blockSize ), mPointCount( pointCount ), mLazInfo( lazInfo )
```
The lazInfo contains the point count in the whole laz file but this point count argument is for how many points in a single block. This is correct :)

NEDJIMAbelgacem committed:
Address reviews + more cleanup

NEDJIMAbelgacem commented on src/core/pointcloud/qgslazdecoder.cpp:
```diff
@@ -400,35 +401,33 @@ QgsPointCloudBlock *QgsLazDecoder::decompressLaz( const QByteArray &byteArrayDat
 
 
-QgsPointCloudBlock *QgsLazDecoder::decompressCopc( const QString &filename, uint64_t blockOffset, uint64_t blockSize, int32_t pointCount, const QgsPointCloudAttributeCollection &attributes, const QgsPointCloudAttributeCollection &requestedAttributes, const QgsVector3D &_scale, const QgsVector3D &_offset, QgsPointCloudExpression &filterExpression )
+QgsPointCloudBlock *QgsLazDecoder::decompressCopc( const QString &filename, QgsLazInfo &lazInfo, uint64_t blockOffset, uint64_t blockSize, int32_t pointCount, const QgsPointCloudAttributeCollection &attributes, const QgsPointCloudAttributeCollection &requestedAttributes, const QgsVector3D &_scale, const QgsVector3D &_offset, QgsPointCloudExpression &filterExpression )
```
Actually we read blocks concurrently so we end up having to open the file multiple times to avoid data races.

NEDJIMAbelgacem committed:
- Remove unnecessary function
- Fix tests

NEDJIMAbelgacem committed:
fix spelling

NEDJIMAbelgacem committed:
fix tests

wonder-sk commented on src/core/pointcloud/qgscopcpointcloudblockrequest.cpp:
```diff
@@ -0,0 +1,79 @@
+  {
+    error = QStringLiteral( "Network request error: %1" ).arg( mTileDownloadManagerReply->errorString() );
+  }
+  mErrorStr = QStringLiteral( "Error loading point tile %1: \"%2\"" ).arg( mNode.toString() ).arg( error );
```
it looks like we're setting the error string even if there was no error?

wonder-sk commented on src/core/pointcloud/qgscopcpointcloudblockrequest.cpp:
```diff
@@ -0,0 +1,79 @@
+      }
+      catch ( std::exception &e )
+      {
+        error = QStringLiteral( "Error while decompressing node %1: %2" ).arg( mNode.toString(), e.what() );
```
probably no need to have `mNode.toString()` in the error here, because it will be added to the error string below

wonder-sk commented on src/core/pointcloud/qgslazinfo.cpp:
```diff
@@ -0,0 +1,305 @@
+  if ( mPointFormat == 4 || mPointFormat == 5 || mPointFormat == 9 )
+  {
+    // TODO: handle the Wave Packets attributes
+  }
```
I would suggest to remove any support for point formats 4,5,9,10 because PDAL does not support full waveform data, so we're not going to support them anytime soon either :-)

We should actually have a check in QgsLazInfo when parsing header - if the point format is not 0,1,2,3,6,7,8 then we should abort (better to fail early than having random bugs due to unsupported data format)

NEDJIMAbelgacem replied:
We can just not support the wave packets handling and keep the formats readable? QGIS will behave like the attributes of wave packets don't exist at all.

wonder-sk commented on src/core/providers/copc/qgscopcprovider.cpp:
```diff
@@ -108,10 +109,9 @@ QVariant QgsCopcProvider::metadataClassStatistic( const QString &attribute, cons
 {
+  mIndex->load( dataSourceUri() );
   if ( mIndex->isValid() )
     return;
```
let's fix this now before we forget about it :)

NEDJIMAbelgacem committed:
Fix EPT test

NEDJIMAbelgacem committed:
Address reviews

NEDJIMAbelgacem committed:
Add missing laz attributes handling

NEDJIMAbelgacem committed:
Use better default naming for remote COPC

NEDJIMAbelgacem committed:
- Add error checking in the point cloud index
- minor fixes

NEDJIMAbelgacem committed:
small fixes

wonder-sk reviewed (approved):
:clap: Well done!
